anndata/_core/views.py

```diff
@@ -273,7 +273,7 @@ def as_view_awkarray(array, view_args):
                 "Please open an issue in the AnnData repo and describe your use-case."
             )
         array = ak.with_parameter(array, _PARAM_NAME, (parent_key, attrname, keys))
-        array = ak.with_parameter(array, "__array__", "AwkwardArrayView")
+        array = ak.with_parameter(array, "__list__", "AwkwardArrayView")
         return array
 
     ak.behavior["AwkwardArrayView"] = AwkwardArrayView
```

Comment:
```suggestion
    ak.behavior["AwkwardArrayView"] = AwkwardArrayView
    ak.behavior["*", "AwkwardArrayView"] = AwkwardArrayView
```

Reply:
> (a) Would `ak.with_parameter(array, "__list__", None)`

No, the same function `with_list` should be used, with a value of `None`.

> (b) What behavior would this change allow? E.g. does __setitem__ even work for the children of this array?

The main change in Awkward is to replace the powerful `__array__` with a much tighter-scoped `__list__` parameter. This means that if your top-level layout is not an `is_list` type, it won't do anything (and in future, would error). The `with_list` function behaves like `with_name`, in that it descends until if finds a list, sets the parameter, and exits. 

Now that I think about it, the most comprehensive solution would set _either_ `__list__` or `__record__`, whichever appears first. That would look like
```python
def with_any_name(array, name):
    def apply(layout, **kwargs):
        if layout.is_list:
            return layout.with_parameter("__list__", name)
        elif layout.is_record:
            return layout.with_parameter("__record__", name)
    return ak.transform(apply, array)
```

Reply:
Hmm, actually, I suppose that you are only worried about mutability from fields (which is the only kind Awkward Arrays support). Therefore, we should just use `with_name` here: if the layout doesn't have any records, it can't be mutated. So that's the proper solution after all.

Reply:
> That said, I suppose that if you only worry about mutability from fields (which is the only kind Awkward Arrays support), then we should just use with_name here: if the layout doesn't have any records, it can't be mutated.

:+1:

Make sense to me. So, something like:

```python
ak.with_name(array, "AwkwardArrayView")
```

instead of:

```python
ak.with_parameter(array, "__list__", "AwkwardArrayView")
```

?

This does seem to be causing some test failures, but I haven't looked too deeply at that yet.

----

Btw, the current approach works even if there are no lists. E.g.:

```python
a = ad.AnnData(
    np.ones((3, 2)),
    obsm={
        "awk": ak.Array(
            [
                {"a": {"b": 1}},
                {"a": {"b": 2}},
                {"a": {"b": 3}}
            ]
        )
    }
)
```

Is that intentional?

Reply:
So, if I do `with_name` I seem to need `ak.behavior["*", "AwkwardArrayView"] = AwkwardArrayView`

Reply:
@agoose77, could you provide an example that would break given the current behavior?

Reply:
Gah, you're quicker than I can type!

> `ak.with_name(array, "AwkwardArrayView")`

Yes, this. Explicitly, what this means is that only arrays with records *somewhere* in the layout will be given your special behavior class. It so happens that only arrays with records can be mutated in-place, though, so that's exactly what we want.

I took a look at `views.py`. We still want to set the `_PARAM_NAME` parameter on the top-level node, because any operation that drops these parameters will involve losing the original `ak.Array` anyway.

Setting the record name means that users won't be able to set their own record names without losing the anndata view. I think that's actually OK - behavior classes are somewhat at odds with anndata's needs here, and I think it's fine for a user to need to write
```python
ak.with_name(anndata['my_vector'], "Vector3D")
```
to get the vector behavior on an anndata-stored Awkward Array. Pinging @jpivarski to have an opinion here, though.

If Jim's happy with this trade-off, then I think this should work. The only thing I also noticed; does the registry store a weakref to the `ak.Array` object? We should test for whether `ak.Array(array)` retains the link to the original view - it shouldn't in my view. Storing the original array in the parameters as a weakref would mean we could check that, and avoid triggering a copy-on-write in such circumstances.

Reply:
> Setting the record name means that users won't be able to set their own record names without losing the anndata view. I think that's actually OK - behavior classes are somewhat at odds with anndata's needs here, and I think it's fine for a user to need to write

This was true before (I think?)

Would it be true with the `ak.with_parameter(array, "__list__", "AwkwardArrayView")` approach?

Playing around a bit:

```python
two = ak.Array([[{"x": 0.9, "y": 1}, {"x": 2, "y": 2.2}, {"x": 2.9, "y": 3}],
                [],
                [{"x": 3.9, "y": 4}, {"x": 5, "y": 5.5}],
                [{"x": 5.9, "y": 6}],
                [{"x": 6.9, "y": 7}, {"x": 8, "y": 8.8}, {"x": 8.9, "y": 9}]],
               with_name="point")

ak.with_name(two, "AwkwardArrayView").typestr
#  '5 * var * AwkwardArrayView[x: float64, y: float64]'

ak.with_parameter(two, "__list__", "AwkwardArrayView").typestr
# '5 * [var * point[x: float64, y: float64], parameters={"__list__": "AwkwardArrayView"}]'
```

I think the `with_parameter` result better fits my mental model of what we're trying to do.

Reply:
`__list__` is defined only for `is_list` types. Here's an example of how fragile that is if we only set the top-most parameter:
```
>>> array = ak.Array([[1,2,3], [4]])
>>> result = array.mask[[True, False]]
>>> ak.with_parameter(result, "__list__", "View").layout
<ByteMaskedArray valid_when='true' len='2'>
    <parameter name='__list__'>'View'</parameter>
    <mask><Index dtype='int8' len='2'>
        [1 0]
    </Index></mask>
    <content><ListOffsetArray len='2'>
        <offsets><Index dtype='int64' len='3'>
            [0 3 4]
        </Index></offsets>
        <content><NumpyArray dtype='int64' len='4'>[1 2 3 4]</NumpyArray></content>
    </ListOffsetArray></content>
</ByteMaskedArray>
```

This currently works, but it won't forever; `__list__` is defined only for `is_list` types (i.e. the `ListOffsetArray` here, not the `ByteMaskedArray`). We're just gradually introducing these rules, so the sanity checks will come next.

But, you can ensure that we only set the parameter on `__list__` nodes using the recursive transform above. The question is what else breaks?

If you support single-records in AnnData, i.e. `ak.record.Record` objects that wrap an array, then these are top-level nodes that do not have `is_list=True`, so these would not be supported. Records _can_ be mutated just like record _arrays_.

Reply:
Just catching up here...

> That said, I suppose that if you only worry about mutability from fields (which is the only kind Awkward Arrays support), then we should just use with_name here: if the layout doesn't have any records, it can't be mutated.

this makes sense to me. 

I understand the final discussion is whether to use `with_name`, or `with_parameter` here?
 * using `with_name` doesn't have any obvious downsides except that the `name` parameter is lost? In that case we'd probably want to update [this code](https://github.com/scverse/anndata/blob/98ba35999276d2bf15fd9b9e2d86f6df87dcc6ab/anndata/_core/views.py#L270-L274) to raise a `NotImplementedError` when a name is detected instead. 
 * Using `with_parameter` is bound to break when there is no `ListType` in the array. Ist this something that can actually happen, as we require at least one dimension to be aligned to AnnData (except for `adata.uns` ofc.). 
 * Any updates on  https://github.com/scikit-hep/awkward/issues/1391#issuecomment-1412297114? Back then we discussed this as the ideal solution once it is ready.